File: queue.py
class Queue:
    def __init__(self):
        self.arr = []
        self.top = -1
        self.tail = -1
    def __str__(self):
        return str(self.arr)

    def pop(self):
        if len(self.arr) == 0:
            return None
        r = self.arr[0]
        self.arr = self.arr[1:]
        return r 

    def push(self, val):
        self.arr.append(val) 


q = Queue()
q.push(1)
q.push(2)
q.push(3)
print(q)
q.pop()
print(q)
q.pop()
print(q)

File: queue.js
function Queue() {
    this.arr = [];
    this.head = -1;
    this.tail = -1;
    this.init = init;
    this.enqueue = enqueue;
    this.dequeue = dequeue;
    this.size = size;
    this.toString = toString;
}

function init() {
    this.arr = [];
    this.head = -1;
    this.tail = -1;
}

function enqueue(e) {
    if (this.head == -1) {
        this.head = 0;
        this.tail = 0;
        this.arr[this.tail] = e;
        return ;
    }
    this.arr[++this.tail] = e ;
}

function dequeue(e) {
    if (this.size() == 0) {
        return NaN;
    }
    r = this.arr[this.head];
    if (this.size() == 1) {
        this.init();
    }else {
        this.arr = this.arr.slice(1,this.size());
    }
    return r;
}

function size() {
    return this.arr.length;
}

function toString() {
    return this.arr.toString();
}

q = new Queue();
q.enqueue(1);
q.enqueue(2);
q.enqueue(3);
q.enqueue(4);
console.log(q.toString());

console.log(q.dequeue());
console.log(q.toString());

console.log(q.dequeue());
console.log(q.toString());

q.init();
console.log(q.toString());
q.dequeue();
